draft-ietf-tls-tls13.md

```diff
@@ -2785,6 +2792,22 @@ certificate_authorities
   the certificate_authorities list is empty, then the client MAY
   send any certificate of the appropriate ClientCertificateType,
   unless there is some external arrangement to the contrary.
+
+certificate_extensions
+: A list of certificate extension OIDs [RFC5280] with their allowed
+  values, represented in DER-encoded format. Some certificate
+  extension OIDs allow multiple values (e.g. Extended Key Usage).
+  If the server has included a non-empty certificate_extensions list,
+  the client certificate MUST contain all of the specified extension
+  OIDs that the client recognizes. For each extension OID recognized
+  by the client, all of the specified values MUST be present in the
+  client certificate. However, the client MUST ignore and skip any
+  unrecognized certificate extension OIDs. If the client has ignored
+  some of the required certificate extension OIDs, and supplied a
+  certificate that does not satisfy the request, the server MAY at
+  its discretion either continue the handshake without client
+  authentication, or respond with a fatal unsupported_certificate
```

Comment:
With this proposal, certs remain opaque to the protocol.

Any OID matching code remains in the PKI library where it belongs.

Cheers,

Andrei

From: Rich Salz [mailto:notifications@github.com]
Sent: Tuesday, August 18, 2015 9:31 AM
To: tlswg/tls13-spec <tls13-spec@noreply.github.com>
Cc: Andrei Popov <Andrei.Popov@microsoft.com>
Subject: Re: [tls13-spec] TLS Client Auth (#209)


In draft-ietf-tls-tls13.md<https://github.com/tlswg/tls13-spec/pull/209#discussion_r37320488>:

> @@ -2785,6 +2792,22 @@ certificate_authorities

>    the certificate_authorities list is empty, then the client MAY

>    send any certificate of the appropriate ClientCertificateType,

>    unless there is some external arrangement to the contrary.

> +

> +certificate_extensions

> +: A list of certificate extension OIDs [RFC5280] with their allowed

> +  values, represented in DER-encoded format. Some certificate

> +  extension OIDs allow multiple values (e.g. Extended Key Usage).

> +  If the server has included a non-empty certificate_extensions list,

> +  the client certificate MUST contain all of the specified extension

> +  OIDs that the client recognizes. For each extension OID recognized

> +  by the client, all of the specified values MUST be present in the

> +  client certificate. However, the client MUST ignore and skip any

> +  unrecognized certificate extension OIDs. If the client has ignored

> +  some of the required certificate extension OIDs, and supplied a

> +  certificate that does not satisfy the request, the server MAY at

> +  its discretion either continue the handshake without client

> +  authentication, or respond with a fatal unsupported_certificate

Right now certs are pretty opaque to the protocol, and many TLS implementations. Understanding X509v3 extensions seems way out of scope for TLS. For example, we don't even talk about validity periods! Strike this and leave it up to the application.

—
Reply to this email directly or view it on GitHub<https://github.com/tlswg/tls13-spec/pull/209/files#r37320488>.